### PR TITLE
PR for #14: Create binned scatter function

### LINKED ISSUE #14: Create binned scatter function
Per comment [here](https://github.com/gslab-econ/ad-price-drivers/issues/503#issuecomment-916384994), goal is to adapt the binned scatter function(s) we use in ad-price-driver for cross-project/repo usage. Relevant code and documentation can be found [here](https://github.com/gslab-econ/ad-price-drivers/tree/ab7ae2f4def6a12261bf384dc473e0dd781a493f/analysis/source/lib/plot) (note not all files in the linked directory is for creating/testing `binscatter.R`).

@jmshapir fyi.

### PR DESCRIPTION
**Pull request by ew487**
Closes #14 

Note: confidence intervals still in progress.

### CODE REVIEW

**Comment by ew487** [comment]
Hi @jmshapir, I'm having trouble adding @torenfronsdal as a reviewer, do you know how to fix this?

File: BinScatter/R/BinScatter.R
Comment (torenfronsdal) [inline BinScatter/R/BinScatter.R:17]:
```suggestion
#' "quantile", default is `binType`).
```

**Comment by jmshapir** [comment]
> Hi @jmshapir, I'm having trouble adding @torenfronsdal as a reviewer, do you know how to fix this?

I added @torenfronsdal to the repository and was able to add @torenfronsdal as a reviewer.

(For now @torenfronsdal has read access; we can elevate to write if/as needed.)

File: BinScatter/R/BinScatter.R
Comment (torenfronsdal) [reply BinScatter/R/BinScatter.R:16]: Could change to `n_bins`, `bin_type`, `n_partial_bins`, `partial_bin_type`, `drop_na`.

File: BinScatter/R/BinScatter.R
Comment (torenfronsdal) [inline BinScatter/R/BinScatter.R:29]:
```suggestion
#' @param plot_xlab X-axis title for the binned scatter plot (string, default is `x_var`).
#' @param plot_ylab Y-axis title for the binned scatter plot (string, default is `y_var`).
```

File: BinScatter/R/BinScatter.R
Comment (torenfronsdal) [inline BinScatter/R/BinScatter.R:169]:
```suggestion
  df_reg <- data %>% dplyr::select(eval(y_var))
```

When I run the example code, I get the error `Error in select(., eval(y_var)) : unused argument (eval(y_var))` because there is a naming conflict for `select` from `dplyr` and `MASS`.

**Comment by ew487** [comment]
> I added @torenfronsdal to the repository and was able to add @torenfronsdal as a reviewer.

Great, thanks @jmshapir @torenfronsdal !

**Comment by ew487** [comment]
Update: I just found a problem with `linpartial_var` & working on debugging.

File: BinScatter/R/BinScatter.R
Comment (torenfronsdal) [inline BinScatter/R/BinScatter.R:180]:
```suggestion
    if (is.null(partialBinType)) { partialBinType <- binType }
```

File: BinScatter/R/BinScatter.R
Comment (torenfronsdal) [inline BinScatter/R/BinScatter.R:196]: For code clarity, maybe these lines should be in an `else` clause for the `if` statement below, since this applies only when `intercept == FALSE`.

File: BinScatter/R/BinScatter.R
Comment (torenfronsdal) [inline BinScatter/R/BinScatter.R:239]:
```suggestion
  if (is.null(plot_xlab)) { plot_xlab <- x_var }
  if (is.null(plot_ylab)) { plot_ylab <- y_var }
```

**Comment by torenfronsdal** [comment]
@ew487 We may want to also add a test case where we implement `linpartial_var` for some sample data.

**Comment by ew487** [comment]
Thanks @torenfronsdal, that's a good idea! I'll work on that.

**Comment by ew487** [comment]
> Update: I just found a problem with `linpartial_var` & working on debugging.

It turns out that the bug does not have to do with `linpartial_var` but actually occurs when some bins have no observations (for example, when I simulate data using low `N` and have high `nBins`). @jmshapir @torenfronsdal Could you please let me know what you prefer between adding code to handle this case, or raising an error & terminating?

**Comment by torenfronsdal** [comment]
> > Update: I just found a problem with `linpartial_var` & working on debugging.
> 
> It turns out that the bug does not have to do with `linpartial_var` but actually occurs when some bins have no observations (for example, when I simulate data using low `N` and have high `nBins`). @jmshapir @torenfronsdal Could you please let me know what you prefer between adding code to handle this case, or raising an error & terminating?

@ew487 @jmshapir Maybe we should decide on a minimum number of obs. per bin and throw an error if we have fewer. 

This [binscatter](https://github.com/maximilianeber/binscatter/blob/04f928d2cab5604e078f45a7eb9e7281f723d2c0/stat_binscatter.R#L5) function (by my former coworker) requires at least 10 obs. per bin.

**Comment by ew487** [comment]
@torenfronsdal Ok! That makes sense. I'll add something like this.
Thanks for this info!

**Comment by jmshapir** [comment]
> > 
> 
> @ew487 @jmshapir Maybe we should decide on a minimum number of obs. per bin and throw an error if we have fewer.
> 
> This [binscatter](https://github.com/maximilianeber/binscatter/blob/04f928d2cab5604e078f45a7eb9e7281f723d2c0/stat_binscatter.R#L5) function (by my former coworker) requires at least 10 obs. per bin.

This sounds fine to me. If we go this route, I think we should set a default minimum but let the user override if they wish.

If the resulting regression model is underidentified I think we should just let R throw whatever is the default error (presumably controlled by the underlying regression command) and terminate.

File: BinScatter/DESCRIPTION
Comment (jmshapir) [inline BinScatter/DESCRIPTION:2]:
```suggestion
Title: Binned Scatterplots with Controls
```

File: BinScatter/LICENSE
Comment (jmshapir) [inline BinScatter/LICENSE:1]:
```suggestion
Copyright (c) 2022 gslab-econ
```

File: BinScatter/R/BinScatter.R
Comment (jmshapir) [inline BinScatter/R/BinScatter.R:257]: This script is a little long. It's fairly clear so I don't think it's a big problem, but if there were two or three more functions we thought could go in their own scripts (or in a helper script), that might be nice.

**Review by jmshapir — CHANGES_REQUESTED** [review]
Thanks @ew487 for great work! Generally I found the structure pretty clear, just had a few comments and questions.

File: BinScatter/R/BinScatter.R
Comment (ew487) [reply BinScatter/R/BinScatter.R]: Fixed!

**Comment by ew487** [comment]
Update: I've added a test including `linpartial_var` in [1a427a7](https://github.com/gslab-econ/gslab_r/pull/22/commits/1a427a746c84e765ab580c75d69b4e746345e0a0), and the minimum observations per bin is implemented in [f83a098](https://github.com/gslab-econ/gslab_r/pull/22/commits/f83a09839a107238d15ebc066527a83b05663cad) (in the `Scatter` function that is renamed to `BinScatter` in [2522030](https://github.com/gslab-econ/gslab_r/pull/22/commits/25220301a58088f291e66ad6a40e32a7dae23902)). Also, the `GetMidpoint`, `GetMeanval`, and `MakeBinIndicator` functions are now standalone instead of locally defined functions in `BinScatter`.

File: BinScatter/R/BinScatter.R
Comment (ew487) [reply BinScatter/R/BinScatter.R]: @jmshapir I took out the functions for getting midpoint/mean value, as well as making bins. You can see the result in the `Scatter.R` file in [0bc0013](https://github.com/gslab-econ/gslab_r/pull/22/commits/0bc0013ec9f9ec7675a505a5464530b3366b09a4). If it is still too long please let me know & I can take out more!

**Review by jmshapir — APPROVED** [review]
Thanks @ew487! All of my original comments are addressed and I didn't spot anything new on another scan, so I'm approving.

@torenfronsdal if you are likewise good here can you please "approve" for bookkeeping purposes? After that @ew487 can merge and we can put this package to use!

Thanks again all.

**Review by torenfronsdal — APPROVED** [review]

**Comment by ew487** [comment]
Summary [here](https://github.com/gslab-econ/gslab_r/issues/14#issuecomment-1224804605).

**Pull request merged**
